src/telemetry/Manager.cc

```diff
@@ -545,6 +575,25 @@ HistogramPtr Manager::HistogramInstance(std::string_view prefix, std::string_vie
     return HistogramInstance(prefix, name, lbls, bounds_span, helptext, unit);
 }
 
+void Manager::Process() {
+    std::unique_lock<std::mutex> lk(collector_cv_mtx);
+
+    collector_flare->Extinguish();
+    prometheus_registry->UpdateViaCallbacks();
+
+    lk.unlock();
+    collector_cv.notify_all();
+}
+
+void Manager::WaitForPrometheusCallbacks() {
+    std::unique_lock<std::mutex> lk(collector_cv_mtx);
+    collector_flare->Fire();
+
+    // It should *not* take 5 seconds to go through all of the callbacks, but
+    // set this to have a timeout anyways just to avoid a deadlock.
+    collector_cv.wait_for(lk, std::chrono::seconds(5));
```

Comment:
Should this handle a timeout return value and print to stderr if there is one? A BifConst variable would be nice for testing really small timeouts to try and force a timeout.

For the latter point, `Manager::Process()` holds the lock while executing callbacks - if that actually takes too long, are we in danger of running into:

> Right after wait_for returns, lock.owns_lock() is true, and lock.mutex() is locked by the calling thread. If these postconditions cannot be satisfied[[1]](https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for#cite_note-1), calls [std::terminate](https://en.cppreference.com/w/cpp/error/terminate).

There's also no predicate we're looping on, so we may run into spurious wakeups, too. Maybe this needs some counter/queue to understand if callback processing has completed? This should also help avoiding to hold the lock in `Manager::Process()`: Only hold it to increase a counter or send something over the queue.

Reply:
See if that matches what you were thinking. The idea of returning early if the flag is set is that we shouldn't bother to run the callbacks a second time if they're already running. Just let the second request wait until it's done and return the same data.

Reply:
Hmm, that's not it.

The `ProcessFd()` runs on in the main-thread through the IO loop sequentially, so the early return should never be hit and is confusing.

The current change doesn't actually wait for the callbacks to complete. Sprinkling a few fprintfs, you get output which shows the callbacks run *after* the thread has fallen through the condition variable:

```
diff --git a/src/telemetry/Manager.cc b/src/telemetry/Manager.cc
index b370db8758..429bc04d37 100644
--- a/src/telemetry/Manager.cc
+++ b/src/telemetry/Manager.cc
@@ -586,22 +586,32 @@ void Manager::ProcessFd(int fd, int flags) {
 
     collector_flag = true;
     collector_flare->Extinguish();
+    std::fprintf(stderr, "running callbacks\n");
     prometheus_registry->UpdateViaCallbacks();
 
     collector_flag = false;
     lk.unlock();
+    std::fprintf(stderr, "notify!\n\n");
     collector_cv.notify_all();
 }
 
 void Manager::WaitForPrometheusCallbacks() {
+    std::fprintf(stderr, "wait for prometheus callbacks\n");
     std::unique_lock<std::mutex> lk(collector_cv_mtx);
     collector_flare->Fire();
 
     // It should *not* take 5 seconds to go through all of the callbacks, but
     // set this to have a timeout anyways just to avoid a deadlock.
-    bool res = collector_cv.wait_for(lk, std::chrono::seconds(5), []() { return ! telemetry_mgr->collector_flag; });
+    bool res = collector_cv.wait_for(lk, std::chrono::seconds(5), []() {
+        bool p = ! telemetry_mgr->collector_flag;
+        std::fprintf(stderr, "pred %d\n", p);
+        return p;
+    });
+
     if ( ! res )
         fprintf(stderr, "Timeout waiting for prometheus callbacks\n");
+
+    std::fprintf(stderr, "wait for prometheus callbacks done\n");
 }
 
 } // namespace zeek::telemetry
```

```
# zeek -C -i lo Telemetry::metrics_port=4222/tcp  Log::default_writer=Log::WRITER_NONE
<params>, line 1: listening on lo

wait for prometheus callbacks
pred 1
wait for prometheus callbacks done
running callbacks
notify!

wait for prometheus callbacks
pred 1
wait for prometheus callbacks done
running callbacks
notify!

wait for prometheus callbacks
pred 1
wait for prometheus callbacks done
running callbacks
notify!

```

Reply:
This is the local patch that seems to work alright with lots of printf'ing.
```
diff --git a/src/telemetry/Manager.cc b/src/telemetry/Manager.cc
index b370db8758..8f7213431b 100644
--- a/src/telemetry/Manager.cc
+++ b/src/telemetry/Manager.cc
@@ -156,11 +156,21 @@ void Manager::InitPostScript() {
 }
 
 void Manager::Terminate() {
+    {
+        // Notify threads that we're shutting down.
+        //
+        // Don't think we really need the lock here?
+        std::unique_lock<std::mutex> lk(collector_cv_mtx);
+        std::fprintf(stderr, "wake em up term=%d\n", zeek::run_state::terminating);
+        collector_cv.notify_all();
+    }
     // Shut down the exposer first of all so we stop getting requests for
     // data. This keeps us from getting a request on another thread while
     // we're shutting down.
+    std::fprintf(stderr, "reset exposer\n");
     prometheus_exposer.reset();
 
+    std::fprintf(stderr, "unregister\n");
     iosource_mgr->UnregisterFd(collector_flare->FD(), this);
     collector_flare.reset();
 }
@@ -579,29 +589,37 @@ HistogramPtr Manager::HistogramInstance(std::string_view prefix, std::string_vie
 }
 
 void Manager::ProcessFd(int fd, int flags) {
-    if ( collector_flag )
-        return;
-
     std::unique_lock<std::mutex> lk(collector_cv_mtx);
-
-    collector_flag = true;
     collector_flare->Extinguish();
+
+    std::fprintf(stderr, "running callbacks\n");
     prometheus_registry->UpdateViaCallbacks();
+    response_idx = request_idx; // Let pending threads know we've completed the update.
+    std::fprintf(stderr, "running callbacks updated to %lu\n", request_idx);
 
-    collector_flag = false;
     lk.unlock();
+    std::fprintf(stderr, "notify!\n");
     collector_cv.notify_all();
 }
 
 void Manager::WaitForPrometheusCallbacks() {
     std::unique_lock<std::mutex> lk(collector_cv_mtx);
+    ++request_idx;
+    uint64_t expected_idx = request_idx;
+    std::fprintf(stderr, "wait for prometheus callbacks %lu\n", expected_idx);
     collector_flare->Fire();
 
     // It should *not* take 5 seconds to go through all of the callbacks, but
     // set this to have a timeout anyways just to avoid a deadlock.
-    bool res = collector_cv.wait_for(lk, std::chrono::seconds(5), []() { return ! telemetry_mgr->collector_flag; });
+    bool res = collector_cv.wait_for(lk, std::chrono::seconds(5), [expected_idx]() {
+        bool p = telemetry_mgr->response_idx >= expected_idx || zeek::run_state::terminating;
+        std::fprintf(stderr, "predicate %d for %lu!\n", p, expected_idx);
+        return p;
+    });
     if ( ! res )
         fprintf(stderr, "Timeout waiting for prometheus callbacks\n");
+
+    std::fprintf(stderr, "wait for prometheus callbacks done idx=%lu\n", expected_idx);
 }
 
 } // namespace zeek::telemetry
diff --git a/src/telemetry/Manager.h b/src/telemetry/Manager.h
index 0627971f0e..3983416e3d 100644
--- a/src/telemetry/Manager.h
+++ b/src/telemetry/Manager.h
@@ -275,7 +275,9 @@ private:
     std::unique_ptr<zeek::detail::Flare> collector_flare;
     std::condition_variable collector_cv;
     std::mutex collector_cv_mtx;
-    bool collector_flag = false;
+    // Only modified under collector_cv_mtx!
+    uint64_t request_idx = 0;
+    uint64_t response_idx = 0;
 };
 
 } // namespace zeek::telemetry

```

Reply:
That commit would add a few BifConst values. Not expecting anyone to change them in, but they should make testing easier for us if there's something going wrong:

https://github.com/zeek/zeek/commit/a7d0d52071114a613672ec37c54718afbe003b6a

Reply:
Ah, ok that makes more sense. I integrated that. Thanks for the BifConst commit. I cherry-pbicked that in too.